engines/bops_core/app/controllers/concerns/bops_core/magic_link_authenticatable.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+module BopsCore
+  module MagicLinkAuthenticatable
+    extend ActiveSupport::Concern
+    SGID_FOR = "magic_link"
+
+    included do
+      rescue_from ActionController::ParameterMissing do |exception|
+        render plain: "Not Found", status: :not_found
+      end
+    end
+
+    def authenticate_with_sgid!
+      resource = GlobalID::Locator.locate_signed(sgid, for: SGID_FOR)
+
+      handle_expired_or_invalid_sgid if resource.nil?
+    end
+
+    private
+
+    def sgid
+      params.require(:sgid)
+    end
+
+    def sgid_expired_resource
+      gid = parse_global_id_from_sgid(sgid)
+      return nil unless gid
+
+      gid.model_class.find_by(id: gid.model_id)
+    rescue ActiveRecord::RecordNotFound
+      nil
+    end
+
+    def handle_expired_or_invalid_sgid
+      expired_resource = sgid_expired_resource
+
+      if expired_resource
+        render plain: "Magic link expired", status: :unprocessable_entity
+      else
+        redirect_failed
+      end
+    end
+
+    def parse_global_id_from_sgid(sgid)
```

Comment:
If link is expired, we want to check that the magic link was previously valid and then extract the consultee resource so we can query with the id and grab the email address to resend the magic link